### PR TITLE
ROM integration

### PR DESCRIPTION
**Pull request by krainboltgreene**
ROM is a rather well designed object mapper that could be an incredibly easy to use model system for clearwater. It'll provide an adapter pattern framework for things like localstorage, indexDB, and general AJAX style requests.

  - [ ] Get rom to run in opal
  - [x] Remove old model system
  - [ ] Replace with ROM backed model
  - [ ] Write documentation for creating new Clearwater::Model objects
  - [ ] Create a localstorage adapter
    - [ ] Document the adapter
  - [ ] Create a simple HTTP JSON API adapter
    - [ ] Document the adapter
  - [ ] Create a Rails API adapter
    - [ ] Document the adapter
  - [ ] Create a fixture adapter
    - [ ] Document the adapter
  - [ ] Rip out old documentation
  - [ ] Write event-oriented mapper

### CODE REVIEW

**Comment by vais** [comment]
@krainboltgreene interesting!

**Comment by jgaskins** [comment]
@krainboltgreene One thing I was thinking about that we use at work a lot is updating models via Pusher (or some other websocket library). Our model collections all listen to various Pusher channels for updates to the models and update them on the fly, which then triggers a rerender in the app.

Is this what you meant by the "event-oriented mapper" in your checklist?

**Comment by krainboltgreene** [comment]
Basically.

**Comment by jgaskins** [comment]
:+1:

File: clearwater.gemspec [hunk 1]
```diff
@@ -20,6 +20,7 @@ Gem::Specification.new do |spec|
 
   spec.add_dependency 'opal', '~> 0.7.0'
   spec.add_dependency 'opal-jquery', '~> 0.3.0'
+  spec.add_runtime_dependency "rom"
```
Comment (Kamilandzia) [inline clearwater.gemspec]: Ok

**Comment by wied03** [comment]
Just now checking out Clearwater, but is it the project's goal to tie you into a particular ORM implementation (with a runtime dependency)?

**Comment by krainboltgreene** [comment]
I think Jamie and I have different ideas about that. ROM.rb is an amazing
state/data store library with one of the best adapter patterns I've seen.
That and it being pure Ruby (thus eligible for opal) makes me very happy.

I think that any frontend framework worth it's salt will have a data
manager.

**Comment by jgaskins** [comment]
> Just now checking out Clearwater, but is it the project's goal to tie you into a particular ORM implementation (with a runtime dependency)?

We're still trying to feel it out. I don't think it's a good idea to force an ORM on someone, but conventions (or "reasonable defaults") are good. Without them, you get a lot of fragmentation, so I think once we get a bit more into it we'll probably have at least a recommendation for what to use.

I'd be really interested in seeing how well ROM works in Opal. I think one of the hardest things to overcome would probably be the need to use async operations. @solnic do you have any input on that?

I've been experimenting with a simple, [Redux](http://rackt.github.io/redux)-style data store (called [GrandCentral](https://github.com/clearwater-rb/grand_central)) and I've had some pretty good results with it. I'm hesitant to say which way I think we should go until we see how well everything works together, but until then I'm still very interested in this PR.

**Comment by solnic** [comment]
@jgaskins please expand on `async operations` as I'm not sure what it means in this context

**Comment by wied03** [comment]
Unless we're talking full stack framework, I wouldn't want a front end framework to get involved w/ my ORM choices. Besides, the problem of front end<->back end communication is so much different than the ORM's target problem (which is back end to DB).

Although I have my questions about Volt, I think they picked their ORM setup precisely because they do **not** want a full ORM running on the Opal side. They only need that on the server side.

**Comment by krainboltgreene** [comment]
Sorry, to be clear, this is a frontend data persistence engine, not the
API's data persistence engine.

On Wed, Sep 16, 2015 at 2:09 PM, Brady Wied <notifications@github.com>
wrote:

> Unless we're talking full stack framework, I wouldn't want a front end
> framework to get involved w/ my ORM choices. Besides, the problem of front
> end<->back end communication is so much different than the ORM's target
> problem (which is back end to DB).
>
> Although I have my questions about Volt, I think they picked their ORM
> setup precisely because they do *not* want a full ORM running on the Opal
> side. They only need that on the server side.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/clearwater-rb/clearwater/pull/9#issuecomment-140895845>
> .
>



-- 
Kurtis Rainbolt-Greene, Hacker
Software Developer
1631 8th St.
New Orleans, LA
70115

**Comment by wied03** [comment]
Ahh ok. Sorry for the confusion.

**Comment by jgaskins** [comment]
@solnic On the front end, when we make a request to the API, we need to provide a callback to execute when the request finishes. AFAIK, ROM queries to the data source are expected to block, right?

**Comment by jgaskins** [comment]
@wied03 Yeah, it's a bit confusing. This is basically an experiment to see if ROM will compile via Opal into JS to run within the browser and work well enough in a front-end context to store the data we fetch from the server.

**Comment by solnic** [comment]
@jgaskins we can make them non-blocking and support callbacks, rom's pretty flexible, specific functionality can vary across adapters

**Comment by krainboltgreene** [comment]
Yeah, I figured it would be at the adapter level. I'm going to have to
write an HTTP (fetch), localstorage, indexDB, and memory adapters.

On Wed, Sep 16, 2015 at 3:49 PM, Jamie Gaskins <notifications@github.com>
wrote:

> @wied03 <https://github.com/wied03> Yeah, it's a bit confusing. This is
> basically an experiment to see if ROM will compile via Opal into JS to run
> within the browser and work well enough in a front-end context to store the
> data we fetch from the server.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/clearwater-rb/clearwater/pull/9#issuecomment-140918092>
> .
>



-- 
Kurtis Rainbolt-Greene, Hacker
Software Developer
1631 8th St.
New Orleans, LA
70115

**Comment by jgaskins** [comment]
@solnic Nice. I didn't know that. I knew it was flexible, but that's impressive. :-)

**Comment by solnic** [comment]
@jgaskins this is exactly why I decided to drop this whole idea of a common abstract interface for all adapters, you are free to implement rom adapter however you like, as long as its relations can return data things will work

**Comment by wied03** [comment]
For the server side case at the top, Might it be easier to use an existing protocol that can serialize predicates rather than writing a fresh adapter? Odata, graphql etc?

**Comment by krainboltgreene** [comment]
@wied03 We'd still need some nice adapters for data that doesn't go to a server.

**Comment by wied03** [comment]
Yep, I was only commenting on the Rails HTTP adapter mentioned above

**Comment by jgaskins** [comment]
Gonna go ahead and close this. I'd love to see ROM used in a Clearwater app, but I think it might be a good idea to determine if a front-end app is a good use case for ROM first. I really appreciate the discussion that went into this, though.

**Pull request closed without merging**